lib/Smime.php

```diff
@@ -123,13 +141,57 @@ public function addPersonalPublicKey($key, $signkey = false)
      */
     public function addPersonalPrivateKey($key, $signkey = false)
     {
+        global $prefs;
+
         $prefName = $signkey ? 'smime_private_sign_key' : 'smime_private_key';
         $val = is_array($key) ? implode('', $key) : $key;
         try {
             $val = HordeString::convertToUtf8($val);
         } catch (Exception $ex) {
         }
-        $GLOBALS['prefs']->setValue($prefName, $val);
+
+        // check if a private key already exists
+        $check  = $prefs->getValue('smime_private_key');
+
+        // it there is a private key, these will be unset first and then the new one will be loaded
+        if (empty($check)) {
+            $GLOBALS['prefs']->setValue($prefName, $val);
+        } else {
+            $this->unsetSmimePersonal();
+            $GLOBALS['prefs']->setValue($prefName, $val);
+        }
+    }
+
+    /**
+     * Adds am extra personal keys to the extra keys table.
+     *
+     * @param string $pref_name To be removed... TODO.
+     * @param string|array $key  The private key to add.
+     * @param string|array $key  The public key to add.
+     */
+    public function addExtraPersonalKeys($private_key, $public_key, $password, $pref_name = 'smime_private_key')
+    {
+        /* Get the user_name  */
+        // TODO: is there a way to only use prefs?
+        $user_name = $GLOBALS['registry']->getAuth();
+
+        // Encrypt the password
+        $key = $GLOBALS['conf']['secret_key'];
+        $blowfish = new Horde_Crypt_Blowfish($key);
+        $encryptedPassword = $blowfish->encrypt($password);
+        $encryptedPassword = base64_encode($encryptedPassword);
+
+
+        if (!empty($public_key) && !empty($private_key) && !empty($encryptedPassword)) {
+            /* Build the SQL query. */
+            $query = 'INSERT INTO imp_smime_extrakeys (pref_name, user_name, private_key, public_key, privatekey_passwd) VALUES (?, ?, ?, ?, ?)';
+            $values = [$pref_name, $user_name, $private_key, $public_key, $encryptedPassword];
+        }
+        try {
+            $this->_db->insert($query, $values);
+        } catch (\Throwable $th) {
+            throw $th;
```

Comment:
This works, but at that point why even have the try/catch block?

Reply:
ok I removed it, i thought, maybe to catch an unexpected error

Reply:
ok. There is no differnce in catching an error an then throwing it and not catching it at all.

That would only make sense if you want to do something difference in the catch block (like logging) or if you wanted to throw a different exception

Reply:
ok thank you!